src/main/scala/zio/prelude/NonEmptySet.scala

```diff
@@ -0,0 +1,180 @@
+package zio.prelude
+
+import zio.NonEmptyChunk
+
+import scala.language.implicitConversions
+
+final class NonEmptySet[A] private (private val set: Set[A]) { self =>
+
+  /** Converts this `NonEmptySet` to a `Set`. */
+  def toSet: Set[A] = set
+
+  /**
+   * Returns an element of this `NonEmptySet` and the remainder, which is a (possibly empty) `Set`.
+   */
+  @inline
+  def destruct: (A, Set[A]) = (set.head, set.tail)
+
+  /**
+   * Converts this `NonEmptySet` to a `NonEmptyChunk`.
+   */
+  def toNonEmptyChunk: NonEmptyChunk[A] = destruct match { case (head, tail) => NonEmptyChunk.fromIterable(head, tail) }
+
+  /**
+   * Converts this `NonEmptySet` to a `NonEmptyList`.
+   */
+  def toNonEmptyList: NonEmptyList[A] = destruct match { case (head, tail) => NonEmptyList.fromIterable(head, tail) }
+
+  /** Creates a new `NonEmptySet` with an additional element, unless the element is
+   *  already present.
+   *
+   *  @param elem the element to be added
+   *  @return a new set that contains all elements of this set and that also
+   *          contains `elem`.
+   */
+  def +(elem: A): NonEmptySet[A] = new NonEmptySet(set + elem)
+
+  /** Computes the union between of `NonEmptySet` and another set.
+   *
+   *  @param   that  the set to form the union with.
+   *  @return  a new `NonEmptySet` consisting of all elements that are in this
+   *  set or in the given set `that`.
+   */
+  def union(that: Set[A]): NonEmptySet[A] = new NonEmptySet(set.union(that))
+
+  /** Creates a new `NonEmptySet` by adding all elements contained in another collection to this `NonEmptySet`, omitting duplicates.
+   *
+   * This method takes a collection of elements and adds all elements, omitting duplicates, into `NonEmptySet`.
+   *
+   * Example:
+   *  {{{
+   *    scala> val a = NonEmptySet(1, 2) ++ NonEmptySet(2, "a")
+   *    a: zio.prelude.NonEmptySet[Any] = NonEmptySet(1, 2, a)
+   *  }}}
+   *
+   *  @param elems     the collection containing the elements to add.
+   *  @return a new `NonEmptySet` with the given elements added, omitting duplicates.
+   */
+  def ++(elems: Iterable[A]): NonEmptySet[A] = new NonEmptySet(set ++ elems)
+
+  /** Adds the `elem` to this `NonEmptySet`. Alias for `+`. */
+  def add(elem: A): NonEmptySet[A] = self + elem
+
+  /** Removes the `elem` from this `NonEmptySet`. Alias for `-`. */
+  def remove(elem: A): Set[A] = set - elem
+
+  /**
+   * Flattens a `NonEmptySet` of `NonEmptySet` values into a single
+   * `NonEmptySet`.
+   */
+  def flatten[B](implicit ev: A <:< NonEmptySet[B]): NonEmptySet[B] =
+    new NonEmptySet[B](set.foldLeft[Set[B]](Set.empty)((b, a) => b union ev(a)))
+
+  override def hashCode: Int = set.hashCode ^ getClass.hashCode
```

Comment:
Shall we extract the NonEmptySet hash code into a `val` so we don't have to recompute it each time we hash something?

Reply:
:heavy_check_mark: